File: QuickSort.py
# encoding:utf-8

a = [12, 3, 5, 8, 10]


def Partition(a, left, right):
    key = a[left]
    while (left < right):
        while (left < right and a[right] >= key):
            right -= 1
        a[left] = a[right]
        while (left < right and a[left] <= key):
            left += 1
        a[right] = a[left]
    a[left] = key
    return left


def qsort(a, left, right):
    if (left < right):
        # mid=(left+right)/2
        mid = Partition(a, left, right)
        qsort(a, left, mid - 1)
        qsort(a, mid + 1, right)


if __name__ == '__main__':
    a = [12, 3, 5, 8, 10]
    qsort(a, 0, 4)
    for i in xrange(0, len(a)):
        print(a[i])

File: quicksort.cpp
int partition(vector<int> &nums, int s, int e)
{
  int tmp = nums[s];
  int i = s, j = e;
  while(i<j)
  {
    while(i<j&&nums[j]>=tmp)
    {
      j--;
    }
    a[i] = a[j];
    while(i<j&&nums[i]<=tmp)
    {
      i++;
    }
    a[j]=a[i];
  }
  a[i] = tmp;
  return i;
}
void quicksort(vector<int> &nums, int s, int e)
{
  if(s>=e)
  {
    return;
  }
  int mid = partition(nums,s,e);
  quicksort(nums, s, mid-1);
  quicksort(nums, mid+1, e);
}
